src/theory/builtin/theory_builtin_rewriter.cpp

```diff
@@ -214,26 +213,57 @@ Node TheoryBuiltinRewriter::getArrayRepresentationForLambdaRec(TNode n,
   std::vector< Node > conds;
   std::vector< Node > vals;
   Node curr = n[1];
-  while( curr.getKind()==kind::ITE || curr.getKind()==kind::EQUAL || curr.getKind()==kind::NOT ){
-    Trace("builtin-rewrite-debug2") << "  process condition : " << curr[0] << std::endl;
+  Kind ck = curr.getKind();
+  NodeManager* nm = NodeManager::currentNM();
+  while (ck == kind::ITE || ck == kind::EQUAL || ck == kind::NOT
+         || ck == kind::BOUND_VARIABLE)
+  {
     Node index_eq;
     Node curr_val;
     Node next;
-    if( curr.getKind()==kind::ITE ){
+    if (ck == kind::ITE)
+    {
+      Trace("builtin-rewrite-debug2")
+          << "  process condition : " << curr[0] << std::endl;
       index_eq = curr[0];
       curr_val = curr[1];
       next = curr[2];
-    }else{
-      bool pol = curr.getKind()!=kind::NOT;
-      //Boolean case, e.g. lambda x. (= x v) is lambda x. (ite (= x v) true false)
-      index_eq = curr.getKind()==kind::NOT ? curr[0] : curr;
-      curr_val = NodeManager::currentNM()->mkConst( pol );
-      next = NodeManager::currentNM()->mkConst( !pol );
     }
-    if( index_eq.getKind()!=kind::EQUAL ){
-      // non-equality condition
-      Trace("builtin-rewrite-debug2") << "  ...non-equality condition." << std::endl;
-      return Node::null();
+    else
+    {
+      Trace("builtin-rewrite-debug2")
+          << "  process base : " << curr << std::endl;
+      // Boolean return case, e.g. lambda x. (= x v) becomes
+      // lambda x. (ite (= x v) true false)
+      index_eq = curr;
+      curr_val = nm->mkConst(true);
+      next = nm->mkConst(false);
+    }
+    if (index_eq.getKind() != kind::EQUAL)
+    {
+      bool pol = index_eq.getKind() != kind::NOT;
+      Node indexEqAtom = pol ? index_eq : index_eq[0];
+      if (indexEqAtom.getKind() == kind::BOUND_VARIABLE)
```

Comment:
You made the control flow of this loop much more complicated... wouldn't it be simpler if you moved this test to beginning? You will might need to recompute a couple things but I think it would be best to not make the whole thing more complicated because of this special case.

Reply:
I don't know how you mean to move this check to the beginning of the loop.  The first step of the loop is to determine `index_eq`, then we we have to check whether its an equality or a variable?
I'll add documentation on what is going on to make this clearer.

Reply:
You could turn it into an equality in the beginning, I mean, so that the rest of the loop can go back to how it was before.

Reply:
There are still two cases: when the Boolean variable is a condition of an ITE, and when it is in the base case position.  So I would have to duplicate the control flow of the block between [1] and [2] in my latest documentation.

Reply:
I see now. Too bad. :(